### PR TITLE
[Bootcamp] - Tochi - Submission for Week 6 Exercise

### PR DESCRIPTION
**Pull request by Tochi-Nwachukwu**
# [Bootcamp] - Tochi - Submission for Week 6 Exercise

## Project Overview

This project fine-tunes OpenAI's GPT-4o-mini model to predict product prices based solely on product descriptions. The goal is to improve upon the baseline mean absolute error of $76 by training the model to estimate prices from textual product information.

The dataset used for this project is the **Amazon Product Price Prediction Dataset** curated specifically for LLM fine-tuning by Jai Keshav Sharma (2024). This dataset contains 400,000 Amazon product listings with detailed descriptions and corresponding prices, making it ideal for training language models on price estimation tasks.

### Dataset Citation
```
@dataset{sharma2024amazon_price_prediction,
  title={Amazon Product Price Prediction Dataset: Curated for LLM Fine-tuning},
  author={Jai Keshav Sharma},
  year={2024},
  publisher={Hugging Face},
  url={https://huggingface.co/datasets/ksharma9719/Amazon-Reviews-2023-curated_for_price_prediction}
}
```

### Test 1: Validation on Training Dataset
When validating the fine-tuned model against a sample of the test dataset, the model achieved strong performance with a mean absolute error of approximately **$28**, significantly improving upon the $76 baseline.

Model performance on training dataset test split

### Test 2: Generalization to External Dataset
When tested against @ed-donner's dataset, the model showed reduced performance with a mean absolute error of approximately **$106.11**, indicating limited generalization to products outside the training distribution.

Model performance on external dataset

## Key Learnings

- **Domain Specificity**: A fine-tuned model can perform exceptionally well on data similar to its training distribution but may struggle when applied to different product categories or pricing structures.

- **Data Quality and Scale**: Careful data cleaning and selecting an appropriately sized training dataset are critical factors in improving model accuracy and prediction reliability.

- **Generalization Challenges**: Fine-tuning on a specific dataset can lead to overfitting, limiting the model's ability to generalize to broader use cases.

### CODE REVIEW

**Comment by ed-donner** [comment]
Hey! Great analysis. Something looks up with the first diagram though - you should never get results like that on validation data. My guess is that you're passing in the price by mistake - can you double check?

**Comment by Tochi-Nwachukwu** [comment]
# [Bootcamp] - Tochi - Week 6 Exercise - Bug Fix for Data Leakage into model

Hi @ed-donner, I have fixed the bug with my solution; the error was from my messages_for() function. The function had a bug, which embedded the price of each product at the end of the description for each training and validation data set. The results look better now

**Screenshot of the performance of the new model**

image

**Comment by ed-donner** [comment]
gotcha - nicely found and fixed!
